source/textInfos/offsets.py

```diff
@@ -221,7 +226,15 @@ def _getFormatFieldAndOffsets(self,offset,formatConfig,calculateOffsets=True):
 		return formatField,(startOffset,endOffset)
 
 	def _getCharacterOffsets(self,offset):
-		return [offset,offset+1]
+		# Windows Unicode is UTF-16, so a character may be two offsets for code points beyond 16 bits.
+		char = self._getTextRange(offset, offset + 1)
+		if HIGH_SURROGATE_FIRST <= char <= HIGH_SURROGATE_LAST:
+			# High (leading) surrogate; next offset is also part of this character.
+			return offset, offset + 2
+		elif offset > 0 and LOW_SURROGATE_FIRST <= char <= LOW_SURROGATE_LAST:
+			# Low (trailing) surrogate; previous offset is also part of this character.
+			return offset - 1, offset + 1

```

Comment:
One thing I didn't consider when I wrote this code is what happens for invalid pairs like trailing then trailing, leading then leading, leading then non-surrogate, etc. This won't cause exceptions, but it seems that most implementations treat them as individual characters in this case. For example, U+d800 U+d800 (��) is treated as two separate characters by both Notepad and Firefox.

I think we can get around this by fetching three offsets: the one before, the current one and the one after. With some cleverness, this shouldn't be too tedious:

```
if offset > 0:
	chars = self._getTextRange(offset - 1, offset + 2)
	# Slicing avoids the need to check length. If invalid, it'll be the empty string.
	prevChar = chars[0:1]
	curChar = chars[1:2]
	nextchar = chars[2:3]
else:
	chars = self._getTextRange(offset, offset + 2)
	prevChar = u""
	# Slicing avoids the need to check length. If invalid, it'll be the empty string.
	curchar = chars[0:1]
	nextChar = chars[1:2]
if HIGH_SURROGATE_FIRST <= curChar <= HIGH_SURROGATE_LAST and LOW_SURROGATE_FIRST <= nextChar <= LOW_SURROGATE_LAST:
... You get the idea.
```

Note that because prevChar will be empty if we're at offset 0, you don't have to check offset > 0 again; any checks against prevChar will be False.